File: AkkaUnitTest/src/test/scala/org/akka/essentials/unittest/example/TickTockTest.scala
package org.akka.essentials.unittest.example

import org.junit.Assert
import com.typesafe.config.ConfigFactory
import akka.actor.ActorSystem
import akka.dispatch.Await
import akka.pattern.ask
import akka.testkit.TestActorRef
import akka.util.duration.intToDurationInt
import akka.util.Timeout
import junit.framework.TestCase
import org.junit.runner.RunWith

class TickTockTest extends TestCase {

  implicit val system = ActorSystem("TestSys", ConfigFactory
    .load().getConfig("TestSys"))

  override def setUp() = {
  }

  override def tearDown() = {
  }

  def testOne() = {
    val actorRef = TestActorRef[TickTock]

    // get access to the underlying actor object
    val actor: TickTock = actorRef.underlyingActor
    // access the methods the actor object and directly pass arguments and
    // test
    actor.tock(new Tock("some message"))
    Assert.assertTrue(actor.state)
  }

  def testTwo() = {
    val actorRef = TestActorRef[TickTock]

    implicit val timeout = Timeout(5 seconds)
    val future = (actorRef ? new Tick("msg")).mapTo[String]
    val result = Await.result(future, timeout.duration)

    Assert.assertEquals("processed the tick message", result)
  }

  def testThree() = {
    val actorRef = TestActorRef[TickTock]

    try {
      actorRef.receive("do something")
      //should not reach here
      Assert.fail()
    } catch {
      case e: IllegalArgumentException => Assert.assertEquals(e.getMessage(), "boom!")
    }
  }
}

File: AkkaUnitTest/src/test/java/org/akka/essentials/unittest/example/TickTockTest.java
package org.akka.essentials.unittest.example;

import static akka.pattern.Patterns.ask;
import junit.framework.Assert;

import org.akka.essentials.unittest.actors.TickTock;
import org.akka.essentials.unittest.actors.TickTock.Tick;
import org.akka.essentials.unittest.actors.TickTock.Tock;
import org.junit.Test;

import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.dispatch.Await;
import akka.testkit.TestActorRef;
import akka.util.Duration;

import com.typesafe.config.ConfigFactory;

public class TickTockTest {
	static ActorSystem _system = ActorSystem.create("TestSys", ConfigFactory
			.load().getConfig("TestSys"));

	@Test
	public void testOne() {
		TestActorRef<TickTock> actorRef = TestActorRef.apply(new Props(
				TickTock.class), _system);

		// get access to the underlying actor object
		TickTock actor = actorRef.underlyingActor();
		// access the methods the actor object and directly pass arguments and
		// test
		actor.tock(new Tock("tock something"));
		Assert.assertTrue(actor.state);

	}

	@Test
	public void testTwo() throws Exception {
		TestActorRef<TickTock> actorRef = TestActorRef.apply(new Props(
				TickTock.class), _system);

		String result = (String) Await.result(ask(actorRef, new Tick("msg"), 5000),
				Duration.parse("5 second"));

		Assert.assertEquals("processed the tick message", result);
	}

	@Test
	public void testThree() throws Exception {
		TestActorRef<TickTock> actorRef = TestActorRef.apply(new Props(
				TickTock.class), _system);
		try {
			actorRef.receive("do something");
			//should not reach here
			Assert.fail();
		} catch (IllegalArgumentException e) {
			Assert.assertEquals(e.getMessage(), "boom!");
		}
	}
	
}
